### ISSUE TITLE
python-igraph segfaults debugger

### ISSUE THREAD

**Issue by yunesj** [igraph/python-igraph#78]
I'm using the master version of python-igraph and the master version of igraph c core.

With `--disable-tls`, my debugger (PyCharm remote debugger) seg faults:

```
ssh://user@hostname:2222/home/user/projects/agca_directed_remote/venv/bin/python -u /home/user/.pycharm_helpers/pydev/pydevd.py --multiproc --qt-support --client '0.0.0.0' --port 56978 --file /home/user/projects/agca_directed_remote/bin/test.py
pydev debugger: process 4279 is connecting

Connected to pydev debugger (build 145.971)
/home/user/projects/agca_directed_remote/etc/user.conf
bash: line 1:  4279 Segmentation fault      (core dumped) /home/user/projects/agca_directed_remote/venv/bin/python -u /home/user/.pycharm_helpers/pydev/pydevd.py --multiproc --qt-support --client '0.0.0.0' --port 56978 --file /home/user/projects/agca_directed_remote/bin/test.py

Process finished with exit code 139
```

Since it doesn't seem to segfault when I'm not debugging, I was thinking it was a multi-threading issue. With `--enable-tls`, I get:

```
    libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I.. -I../include -I../include -I../src/CHOLMOD/Include -I../src/CHOLMOD/Include -I../src/AMD/Include -I../src/AMD/Include -I../src/COLAMD/Include -I../src/COLAMD/Include -I../src/SuiteSparse_config -I../src/SuiteSparse_config -DNPARTITION -DNTIMER -DNCAMD -Wall -I../optional/glpk -I../src/prpack -DPRPACK_IGRAPH_SUPPORT -fPIC -I/usr/include/libxml2 -MT libigraph_la-igraph_cliquer.lo -MD -MP -MF .deps/libigraph_la-igraph_cliquer.Tpo -c igraph_cliquer.c  -fPIC -DPIC -o .libs/libigraph_la-igraph_cliquer.o
    igraph_cliquer.c:32:1: error: '__thread' before 'static'
     IGRAPH_THREAD_LOCAL static int cliquer_interrupted;
     ^
```

I'm installing it via a pip requirements file (`git+ssh://git@github.com/yunesj/python-igraph.git@master --install-option="--c-core-url=https://github.com/igraph/igraph/archive/master.tar.gz"`).

According to igraph's config.log:

```
configure:3262: checking for gcc
configure:3278: found /usr/bin/gcc
configure:3289: result: gcc
configure:3518: checking for C compiler version
configure:3527: gcc --version >&5
gcc (Ubuntu 4.8.4-2ubuntu1~14.04.3) 4.8.4
```

Any suggestions?

Thanks!

**Comment by yunesj** [comment]
Based on [this](https://gcc.gnu.org/ml/gcc-patches/2013-11/msg01302.html), I tried using `_Thread_local` instead of `__thread` with gcc-4.9/gxx-4.9, but it didn't work.

**Comment by ntamas** [comment]
It looks like `gcc` does not like `__thread` appearing before `static` but it compiles just fine if I swap the order of `IGRAPH_THREAD_LOCAL` and `static` in `src/igraph_cliquer.c` and `src/cliquer/cliquer.c`; see 6e492cb . Can you try recompiling after having applied this commit to see whether the original error you are getting is related to thread-safety or not?

**Comment by yunesj** [comment]
It compiled fine (thanks!), but it's still crashing the debugger. Here's a stack trace:

```
Program received signal SIGSEGV, Segmentation fault.
0x00007f7c42453477 in igraphmodule_Graph_vertex_attributes (self=0x7f7c172a6af8)
    at src/graphobject.c:9325
9325      return PyDict_Keys(ATTR_STRUCT_DICT(&self->g)[ATTRHASH_IDX_VERTEX]);
```
```
(gdb) where
#0  0x00007f7c42453477 in igraphmodule_Graph_vertex_attributes (self=0x7f7c172a6af8)
    at src/graphobject.c:9325
#1  0x000000000049a3b5 in PyEval_EvalFrameEx ()
#2  0x00000000004a090c in PyEval_EvalCodeEx ()
#3  0x000000000049ab45 in PyEval_EvalFrameEx ()
#4  0x00000000004a1c9a in ?? ()
#5  0x00000000004dfe94 in ?? ()
#6  0x0000000000505f96 in PyObject_Call ()
#7  0x00000000004de41a in ?? ()
#8  0x00000000005039eb in ?? ()
#9  0x0000000000505f96 in PyObject_Call ()
#10 0x000000000049b07a in PyEval_EvalFrameEx ()
#11 0x00000000004a1c9a in ?? ()
#12 0x0000000000505f96 in PyObject_Call ()
#13 0x000000000049b07a in PyEval_EvalFrameEx ()
#14 0x00000000004a1c9a in ?? ()
#15 0x00000000004dfe94 in ?? ()
#16 0x00000000004dca56 in PyEval_CallObjectWithKeywords ()
#17 0x000000000059bc99 in ?? ()
#18 0x00000000004e5d0c in _PyObject_Str ()
#19 0x0000000000496ad1 in PyString_Format ()
#20 0x000000000049ae20 in PyEval_EvalFrameEx ()
#21 0x00000000004a090c in PyEval_EvalCodeEx ()
#22 0x0000000000499a52 in PyEval_EvalFrameEx ()
#23 0x0000000000499ef2 in PyEval_EvalFrameEx ()
#24 0x0000000000499ef2 in PyEval_EvalFrameEx ()
#25 0x0000000000499ef2 in PyEval_EvalFrameEx ()
#26 0x00000000004a1c9a in ?? ()
#27 0x0000000000505f96 in PyObject_Call ()
#28 0x000000000049b07a in PyEval_EvalFrameEx ()
#29 0x00000000004a090c in PyEval_EvalCodeEx ()
#30 0x000000000049ab45 in PyEval_EvalFrameEx ()
#31 0x00000000004a1c9a in ?? ()
#32 0x00000000004dfe94 in ?? ()
#33 0x00000000004aaf4c in ?? ()
#34 0x000000000049c26a in PyEval_EvalFrameEx ()
#35 0x00000000004a1c9a in ?? ()
#36 0x00000000004dfe94 in ?? ()
#37 0x0000000000505f96 in PyObject_Call ()
#38 0x00000000004de41a in ?? ()
#39 0x00000000005039eb in ?? ()
#40 0x0000000000499be5 in PyEval_EvalFrameEx ()
#41 0x00000000004a090c in PyEval_EvalCodeEx ()
#42 0x000000000049ab45 in PyEval_EvalFrameEx ()
#43 0x00000000004a090c in PyEval_EvalCodeEx ()
#44 0x000000000049ab45 in PyEval_EvalFrameEx ()
#45 0x00000000004a090c in PyEval_EvalCodeEx ()
#46 0x000000000049ab45 in PyEval_EvalFrameEx ()
#47 0x00000000004a1634 in ?? ()
#48 0x000000000044e4a5 in PyRun_FileExFlags ()
#49 0x000000000044e6a2 in ?? ()
#50 0x000000000049968d in PyEval_EvalFrameEx ()
#51 0x00000000004a090c in PyEval_EvalCodeEx ()
#52 0x0000000000499a52 in PyEval_EvalFrameEx ()
#53 0x00000000004a1634 in ?? ()
#54 0x000000000044e4a5 in PyRun_FileExFlags ()
#55 0x000000000044ec9f in PyRun_SimpleFileExFlags ()
#56 0x000000000044f904 in Py_Main ()
#57 0x00007f7c465b7ec5 in __libc_start_main (main=0x44f9c2 <main>, argc=11, 
    argv=0x7ffdacd8c0f8, init=<optimized out>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7ffdacd8c0e8) at libc-start.c:287
#58 0x0000000000578c4e in _start ()
```

The code that's causing the problem is:
```
import igraph

class ProteinNetwork(igraph.Graph):
    def __init__(self):
        super(ProteinNetwork, self).__init__()

        pass

...
from module import ProteinNetwork
def test0():
    g = ProteinNetwork()
    print g
```

Which runs fine:
```
python -u /home/user/projects/my_proj/bin/test.py
IGRAPH U--- 0 0 --

Process finished with exit code 0

```
but segfaults when debugging:
```
bash: line 1:   848 Segmentation fault      (core dumped)
/home/user/projects/my_proj/venv/bin/python -u /home/user/.pycharm_helpers/pydev/pydevd.py --multiproc --qt-support --client '0.0.0.0' --port 39483 --file /home/user/projects/my_proj/bin/test.py
Process finished with exit code 139

```

**Comment by yunesj** [comment]
I didn't realize it was actually the python-igraph c code rather than the igraph core c code.

**Comment by yunesj** [comment]
The code suggests I can workaround this problem by not inspecting the object before it is fully initialized. Indeed, for the following code:
```
1) class ProteinNetwork(igraph.Graph):
2)     def __init__(self):
3)         super(ProteinNetwork, self).__init__()
4)         pass
```
setting a breakpoint on line 3 causes a segfault, but creating a breakpoint on line 4 is fine.

**Comment by ntamas** [comment]
Nice catch - I think I know where the problem is. I will keep you posted.

**Comment by ntamas** [comment]
Could have been related to #97 which I have fixed recently in  161f39f. I could not reproduce it now so I'm closing the issue; feel free to reopen if it is still there with the most recent master revision.
